added lepton 3DIP and pre-computed muon selectors

gkrintir opened the pull request:
Describe the Pull-Request:
-3DIP variables for leptons useful to discriminate prompt vs nonprompt
-muon selectors are strongly recommended from MUO POG: https://twiki.cern.ch/twiki/bin/view/CMS/SWGuideMuonIdRun2#Muon_selectors_Since_9_4_X

Does the PR pass the test script located at
HeavyIonsAnalysis/JetAnalysis/test/runtest.sh
[X] Yes
[] No

Please make sure to mention (using the @ syntax) anyone who might be interested in this PR.
@bi-ran

gkrintir committed:
added lepton 3DIP and pre-computed muon selectors

bi-ran commented:
hi @gkrintir thanks for the PR! just some minor comments:

for consistency between the 2d and 3d impact parameters, could you please also add a check for `!vertex->isFake()` (like at [here](https://github.com/CmsHI/cmssw/blob/forest_CMSSW_10_3_1/HeavyIonsAnalysis/PhotonAnalysis/plugins/ggHiNtuplizer.cc#L711-L715).)

does `front()` work if the vertex collection is empty? it might be good to check for that and have a default of (0, 0, 0) like the 2d variables.

it would also be nice to move the vertex and transient track builder parts
https://github.com/CmsHI/cmssw/blob/64bdf36d6561a81898094ecef93dbdb59a286076/HeavyIonsAnalysis/PhotonAnalysis/plugins/ggHiNtuplizer.cc#L1047-L1053 out of the ele/mu loops

thanks!

gkrintir commented:
thanks @bi-ran, these are indeed valid points. 

So what I did in my local setup for those cases:

-  if the vertex collection is empty or all vertices are fake I skip the event; assigning a real vertex with (0,0,0) could be avoided instead. This means after [L746](https://github.com/CmsHI/cmssw/blob/64bdf36d6561a81898094ecef93dbdb59a286076/HeavyIonsAnalysis/PhotonAnalysis/plugins/ggHiNtuplizer.cc#L746 ) we could add 

`if (vtxHandle->empty() || (pv.x()==0 && pv.y()==0 && pv.z()==0) )
    return; // skip the event if no or nonfake PV found   `

- Initially I also thought that it would be better not to have the vertex and transient track builder parts in the function definitions. But I saw this is done for other collections too so it didn't break the logic of the code. If you still want outside the function I'd go with:

1. defining trackBuilder and PV  (reco::Vertex&) as global variables. This btw means we can simplify the function call for fill*(), *=electrons, photons, and muons. The code has to be updated to take care of math::XYZPoint  --> reco::Vertex& transition.

2. reading the PV (reco::Vertex&) and passing it as additional argument to functions fill*(), *=electrons, photons, and muons. We keep the definition of trackBuilder in the functions.

3. reading the PV (reco::Vertex&) and trackBuilder and passing them as additional arguments to functions fill*(), *=electrons, photons, and muons

Let me know what you think so that I proceed with the corresponding commit.

bi-ran commented:
@gkrintir for the empty vertex case, i would prefer not to skip the event because this is not done in other trees in the forest and will cause events in different trees to be out of sync.

i do agree that it's not good to assign a real (0,0,0) vertex though.. typically the event selection uses the filters from the hiEvtAnalyzer tree, so we don't accidentally include these events, but perhaps assigning (0,0,-99) might be safer? what do you think?

for the transient track builder part, i think it would be good to have it where the other `ESHandle` are ([here](https://github.com/CmsHI/cmssw/blob/64bdf36d6561a81898094ecef93dbdb59a286076/HeavyIonsAnalysis/PhotonAnalysis/plugins/ggHiNtuplizer.cc#L748-L757)), so (1) sounds good, but i don't mind if you want to pass the reco::Vertex& as an additional argument.

thank you for making the changes!

gkrintir committed:
added extra protections

gkrintir commented:
hi @bi-ran 

checking again the [IPTools::absoluteImpactParameter3D()](https://cmssdt.cern.ch/lxr/source/TrackingTools/IPTools/interface/IPTools.h?v=CMSSW_10_3_3_patch1) initializer I can only pass a reco::Vertex&, and not a pointer that I could have used as a global variable. 

But thinking twice I also think none of my two other options seem nice in terms of coding since we'll have both math::XYZPoint& and reco::Vertex& defined in the analyze() function, meaning:

- if there is no good vertex then we are passing a non-initialized reco::Vertex& 

- passing two arguments related to the same object (PV) but with different types (math::XYZPoint& and reco::Vertex&) is also strange 

- even if we had used a global PV (reco::Vertex*) variable then it would be strange not to use it in other instances like [L1006](https://github.com/CmsHI/cmssw/blob/forest_CMSSW_10_3_1/HeavyIonsAnalysis/PhotonAnalysis/plugins/ggHiNtuplizer.cc#L1006)

So to me it looks like more practical to have these parts in the fill*(),*=electrons,muons despite some code duplication.

Btw I added two additional protections:

1. if (!vtxHandle->empty()) --> the 3DIP variables are initialized to -999

2. if the positions of reco::Vertex& and math::XYZPoint& do not agree, i.e., in the very rare case of having a fake leading-pT vertex, the 3DIP variables are also initialized to -999

Also I changed math::XYZPoint pv(0, 0, 0) --> math::XYZPoint pv(0, 0, -999) as agreed. I chose '-999' and not '-99' following the majority of other instances in the code (35 with '-999' vs 13 with '-99').

But anyway if you still don't agree I can drop the IP3D variables and simply commit the mubits. ;D

Alternatively we could think of substituting math::XYZPoint& --> reco::Vertex&, which is btw more natural but as you said this would require more work and not in synch with what is done in other trees

bi-ran committed:
make trackbuilder a class member, use same vertex

pv -> reco::Vertex, pass reco::Vertex into fill_* functions and use
reco::Vertex::position() to get the coordinates

gkrintir committed:
Merge pull request #3 from bi-ran/gkrintir-forest_CMSSW_10_3_1_leptons

make trackbuilder a class member, use same vertex

gkrintir commented:
thx to @bi-ran we now have one object for PV throughout the class, i.e., reco::Vertex&

this PR is now clean and neat and merged
